### PR TITLE
[fix][java-client] Fix message publishing stuck when enabling batch

### LINKED ISSUE #14864: [Bug][Java-Client] Message publishing is stuck
**Describe the bug**

If the producer enables batch,  the message will add to the batch container first, when the broker state is not `ready`, there is no subsequent logic to send messages in the batch container.

Before PR #14185 there is a timer task to call the method `batchMessageContainer` to send messages in the batch container. The PR #14185 optimize batch flush logic, remove the timer task, add a scheduled task after adding messages.

**Some Codes**

Code block in method `serializeAndSendMessage`
```
boolean isBatchFull = batchMessageContainer.add(msg, callback);
lastSendFuture = callback.getFuture();
payload.release();
if (isBatchFull) {
    batchMessageAndSend(false);
} else {
    maybeScheduleBatchFlushTask();
}
```

If the producer state is not ready, the batch flush task will not be scheduled.
```
private void maybeScheduleBatchFlushTask() {
    if (this.batchFlushTask != null || getState() != State.Ready) {
        log.info("xxxx [{}] MaybeScheduleBatchFlushTask state: {}", topic, getState());
        return;
    }
    scheduleBatchFlushTask(conf.getBatchingMaxPublishDelayMicros());
}
```

**To Reproduce**
1. Create a producer
2. Mock broker exception to close connection.
3. Send normal messages in an async way and call get method or send messages with the transaction.
6. The process will be stuck.

**Expected behavior**
Messages could be published normally.

### PR DESCRIPTION
**Pull request by gaoran10**
Fixes #14864 

### Motivation

Before PR #14185 there is a timer task to call the method `batchMessageContainer` to send messages in the batch container. The PR #14185 optimize batch flush logic, remove the timer task, add a scheduled task after adding messages.

If the producer enables batch,  the message will add to the batch container first, when the broker state is not `ready`, there is no subsequent logic to send messages in the batch container.

**Some Codes**

Code block in method `serializeAndSendMessage`
```
boolean isBatchFull = batchMessageContainer.add(msg, callback);
lastSendFuture = callback.getFuture();
payload.release();
if (isBatchFull) {
    batchMessageAndSend(false);
} else {
    maybeScheduleBatchFlushTask();
}
```

If the producer state is not ready, the batch flush task will not be scheduled.
```
private void maybeScheduleBatchFlushTask() {
    if (this.batchFlushTask != null || getState() != State.Ready) {
        log.info("xxxx [{}] MaybeScheduleBatchFlushTask state: {}", topic, getState());
        return;
    }
    scheduleBatchFlushTask(conf.getBatchingMaxPublishDelayMicros());
}
```

### Modifications

Call method `batchMessageAndSend` to make sure messages in the batch container execute send process after producer re-connect.

### Verifying this change

Add new unit test.

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API: (no)
  - The schema: (no)
  - The default values of configurations: (no)
  - The wire protocol: (no)
  - The rest endpoints: (no)
  - The admin cli options: (no)
  - Anything that affects deployment: (no)

### Documentation

Check the box below or label this PR directly (if you have committer privilege).

Need to update docs? 

- [ ] `doc-required` 
  
  (If you need help on updating docs, create a doc issue)
  
- [x] `no-need-doc` 
  
  (Please explain why)
  
- [ ] `doc` 
  
  (If this PR contains doc changes)

### CODE REVIEW

File: pulsar-broker/src/test/java/org/apache/pulsar/client/impl/ProducerCloseTest.java [hunk 1]
```diff
@@ -99,6 +103,25 @@ public void testProducerCloseFailsPendingBatchWhenPreviousStateNotReadyCallback(
         }
     }
 
+    @Test(timeOut = 10_000)
+    public void brokerCloseTest() throws Exception {
+        initClient();
```
Comment (codelipenghui) [inline pulsar-broker/src/test/java/org/apache/pulsar/client/impl/ProducerCloseTest.java:108]: The client will be created before executing the method, we don't need to init it again.
Not sure why all the tests in this class call initClient(), this is a useless call and the old client
will not be closed. We should remove all the `initClient()` calls for this method.

File: pulsar-broker/src/test/java/org/apache/pulsar/client/impl/ProducerCloseTest.java [hunk 2]
```diff
@@ -99,6 +103,25 @@ public void testProducerCloseFailsPendingBatchWhenPreviousStateNotReadyCallback(
...
+        ProducerImpl<byte[]> producer = (ProducerImpl<byte[]>) pulsarClient.newProducer()
+                .topic(topic)
+                .create();
+        producer.newMessage().value("test".getBytes()).send();
+
+        Optional<Topic> topicOptional = pulsar.getBrokerService()
+                .getTopicReference(TopicName.get(topic).getPartitionedTopicName());
+        Assert.assertTrue(topicOptional.isPresent());
+        topicOptional.get().close(true).get();
+        Assert.assertEquals(producer.getState(), HandlerState.State.Connecting);
```
Comment (codelipenghui) [inline pulsar-broker/src/test/java/org/apache/pulsar/client/impl/ProducerCloseTest.java:121]: I think here will introduce a flaky test? the producer might reconnect to the broker.

**Review by codelipenghui — COMMENTED** [review]
Nice catch!

**Review by michaeljmarshall — COMMENTED** [review]
Great catch @gaoran10! I missed this case when I contributed this feature.

> If the producer enables batch, the message will add to the batch container first, when the broker state is not ready, there is no subsequent logic to send messages in the batch container.

I don't think this is completely correct. The timer is scheduled when `pendingMessages.messagesCount()` returns a non 0 result.

https://github.com/apache/pulsar/blob/69d22112066991b98d4f5f8d20b5e28040628979/pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java#L2195-L2199

The problem only occurs if there are no messages in the `pendingMessages` queue. For example, this can happen if some messages are added to the `batchMessageContainer` during a time when the producer is not `Ready`, but not enough to trigger a batch flush.

Part of the issue is how we count "pendingMessages". One alternative solution is to replace `pendingMessages.messagesCount()` with `getPendingQueueSize()`. However, I don't think we should do that because the messages in the batch container are not actually being "re-sent". These messages have never been sent to the broker.

As I propose in my comment, I think we should consider pushing the logic into the `if (messagesToResend == 0)` conditional block. Let me know what you think.

File: pulsar-broker/src/test/java/org/apache/pulsar/client/impl/ProducerCloseTest.java (hunk 1 above)
Comment (gaoran10) [reply pulsar-broker/src/test/java/org/apache/pulsar/client/impl/ProducerCloseTest.java:108]: OK

File: pulsar-broker/src/test/java/org/apache/pulsar/client/impl/ProducerCloseTest.java (hunk 2 above)
Comment (gaoran10) [reply pulsar-broker/src/test/java/org/apache/pulsar/client/impl/ProducerCloseTest.java:121]: I'll new a long backoff client to make the producer connection don't re-connect immediately, we can find this log, the connection will try again in 5.0 s.

```
2022-03-26T11:21:44,782+0800 [pulsar-client-io-44-1] INFO  org.apache.pulsar.client.impl.ConnectionHandler - [broker-close-test-qAAkx] [test-0-0] Closed connection [id: 0x07f473b5, L:/127.0.0.1:49280 - R:localhost/127.0.0.1:49272] -- Will try again in 5.0 s
```

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java [hunk 3]
```diff
@@ -1802,6 +1802,10 @@ private void resendMessages(ClientCnx cnx, long expectedEpoch) {
                     cnx.channel().close();
                     return;
                 }
+
+                if (isBatchMessagingEnabled()) {
+                    batchMessageAndSend(false);
+                }
```
[the remark this answers is not in the document]
Comment (gaoran10) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java:1808]: That makes sense,  thanks. I add the enable batch check in the method `scheduleBatchFlushTask`, or if the producer didn't enable batch will cause an NPE problem when calling method `batchMessageAndSend`.

**Review by codelipenghui — APPROVED** [review]

**Review by Technoboy- — APPROVED** [review]

**Review by Jason918 — APPROVED** [review]

**Review by aloyszhang — APPROVED** [review]

**Review by RobertIndie — APPROVED** [review]

**Review by michaeljmarshall — APPROVED** [review]
LGTM

File: pulsar-broker/src/test/java/org/apache/pulsar/client/impl/ProducerCloseTest.java (hunk 2 above)
Comment (lhotari) [reply pulsar-broker/src/test/java/org/apache/pulsar/client/impl/ProducerCloseTest.java:140]: Indeed, this introduced a new flaky test, #15277 .

**Comment by lhotari** [comment]
this introduced a new flaky test, #15277 . @gaoran10 please take a look

**Pull request merged**